core/pkg/artifacts/saver.go

```diff
@@ -23,7 +24,80 @@ import (
 	"github.com/wandb/wandb/core/pkg/utils"
 )
 
+// ArtifactSaver manages artifact uploads.
 type ArtifactSaver struct {
+	logger              *observability.CoreLogger
+	graphqlClient       graphql.Client
+	fileTransferManager filetransfer.FileTransferManager
+	fileCache           Cache
+
+	wg sync.WaitGroup
+}
+
+func NewArtifactSaver(
+	logger *observability.CoreLogger,
+	graphqlClient graphql.Client,
+	fileTransferManager filetransfer.FileTransferManager,
+) *ArtifactSaver {
+	return &ArtifactSaver{
+		logger:              logger,
+		graphqlClient:       graphqlClient,
+		fileTransferManager: fileTransferManager,
+		fileCache:           NewFileCache(UserCacheDir()),
+	}
+}
+
+type ArtifactSaveResult struct {
+	ArtifactID string
+	Err        error
+}
+
+// Save asynchronously uploads an artifact.
+//
+// This returns a channel to which the operation result is eventually
+// pushed. The channel is guaranteed to receive exactly one value and
+// then close.
+func (as *ArtifactSaver) Save(
+	ctx context.Context,
+	artifact *spb.ArtifactRecord,
+	historyStep int64,
+	stagingDir string,
+) chan ArtifactSaveResult {
+	resultChan := make(chan ArtifactSaveResult, 1)
+	saveCtx := &ArtifactSaveContext{
+		ctx:                 ctx,
+		logger:              as.logger,
+		graphqlClient:       as.graphqlClient,
+		fileTransferManager: as.fileTransferManager,
+		fileCache:           as.fileCache,
+		artifact:            artifact,
+		historyStep:         historyStep,
+		stagingDir:          stagingDir,
+		maxActiveBatches:    5,
+	}
+
+	as.wg.Add(1)
+	go func() {
+		defer as.wg.Done()
+		artifactID, err := saveCtx.Save()
+
+		resultChan <- ArtifactSaveResult{
+			ArtifactID: artifactID,
+			Err:        err,
+		}
+		close(resultChan)
+	}()
+
+	return resultChan
+}
+
+// Finish blocks until all upload operations complete.
+func (as *ArtifactSaver) Finish() {
+	as.wg.Wait()
```

Comment:
this should probably prevent future calls to `Save`

Reply:
removed it!